New filters

djwf opened the pull request:
I'm putting all the filters I create from the list in issue #11 here, as they are part of the standard, and the filter in pull request #73 is not.

Let me know if I should merge these two or split all filters into their own pull requests.  Also, note that I'll use this to add more filters, and will leave a note when I've finished adding filters.  If you'd prefer I not work that way, let me know and I'll change it up.

djwf commented:
Whoops!  My apologies: I need to rebase this off of master, rather than off of my other pull request.  I'll do that after I get some sleep.

djwf commented:
@johannhof: I am confused as to which tests would go in the `src/filters.rs` file, and which would go in `tests/filters.rs`.  Can you give me some guidance on that?

djwf commented:
@johannhof: I'm working on `truncate`, and think liquid-rust would be well served by using proper unicode parsing of strings, thus grapheme clusters rather than characters.  May I add a dependency to an [external crate for unicode segmentation](https://kbknapp.github.io/clap-rs/unicode_segmentation/index.html) to `Cargo.toml/Cargo.lock`?  Or would you rather that happen in a separate pull request with only the filter that requires that change?

johannhof commented:
@djwf Thanks so much for your work on this. Just wanted to let you know it's not been forgotten, I'm simply traveling between continents for work. I'll get to reviewing this today or tomorrow. :)

djwf commented:
@johannhof No worries at all: I'll just keep plugging away at them.  Note that I'm using the list of filters at the [liquid markup web site](https://shopify.github.io/liquid/) as the authoritative list, as I noticed some of those missing from your list in #11 .

johannhof commented:
Ok, looking at this now. The re-ordering makes it a bit hard to see what exactly changed, but I'll manage. As to your questions:

- `src/filters.rs` is probably the best place to put tests. `tests/filters.rs` is intended to test cases where you want the code to go through the whole lexer, parser, etc. before running the filter (although it includes "normal" tests because I was sloppy with review)
- grapheme clusters: without having read the code yet, I agree that this is a good idea in general. I'm not sure how the Ruby version does it (as we'd ideally want to produce the same output at the end of they day), but I guess that shouldn't stop us from doing it the right way. I wonder if there are places in the lexer that could benefit from this, too. But that's irrelevant to this already pretty big PR.

Cheers

johannhof commented on src/filters.rs:
```diff
@@ -271,47 +277,191 @@ pub fn last(input: &Value, _args: &[Value]) -> FilterResult {
+    try!(check_args_len(args, 0));
+    match *input {
+        Value::Array(ref array) => {
+            let mut reversed = array.clone();
```
Note to self: Look if we can reduce allocation in filters (by modifying `output.rs`) before 1.0

johannhof commented on src/filters.rs:
```diff
@@ -497,6 +567,81 @@ pub fn strip_html(input: &Value, _args: &[Value]) -> FilterResult {
+    let ellipsis = "...".to_string();
+    let append = match args.get(1) {
+        Some(&Str(ref x)) => x,
+        _ => &ellipsis,
```
You can replace this with just `"..."`

johannhof replied:
Ah damn I just noticed you took that from `truncatewords`. Can you please fix this up in `truncatewords`, too? Thank you!

djwf replied:
Yep: fixed both truncate and truncatewords.

djwf commented on src/filters.rs:
```diff
@@ -670,6 +670,23 @@ pub fn truncatewords(input: &Value, args: &[Value]) -> FilterResult {
     }
 }
 
+/// Removes any duplicate elements in an array.
```
Got it, done.

djwf commented on src/filters.rs:
```diff
@@ -271,47 +277,191 @@ pub fn last(input: &Value, _args: &[Value]) -> FilterResult {
+pub fn replace_first(input: &Value, args: &[Value]) -> FilterResult {
+    if args.len() != 2 {
+        return Err(InvalidArgumentCount(format!("expected 2, {} given", args.len())));
+    }
```
Got it, done.

djwf commented on src/template.rs:
```diff
@@ -1,10 +1,44 @@
 use context::Context;
-use filters::{size, upcase, downcase, capitalize, minus, plus, times, divided_by, ceil, floor,
-              round, prepend, append, first, last, pluralize, replace_first, replace, date, sort,
-              slice, modulo, escape, escape_once, remove_first, remove, strip_html, truncatewords};
```
That was just because it makes it match the list of filters in the `impl Renderable ...` section, and it made it easier to see where/what was added.  I don't have strong feelings about it either way, so I'll put it back to a grouped version (ordered alphabetically).

djwf commented:
Thanks for the answers and the review: note that the most recent nightly of `clippy` is complaining about the five uses of `lazy_static!` (four in lexer.rs, one in filters.rs) so (on my end, at least) the `rustup run nightly cargo clippy` fails.

Also, my rustfmt (`0.7.1 ()`) formats the grouped filter use statement as follows:

    use filters::{abs, append, capitalize, ceil, date, divided_by, downcase, escape, escape_once, first,
                  floor, join, last, lstrip, minus, modulo, newline_to_br, pluralize, plus, prepend,
                  remove, remove_first, replace, replace_first, reverse, round, rstrip, size, slice,
                  sort, split, strip, strip_html, strip_newlines, times, truncate, truncatewords, uniq,
                  upcase};

The rustfmt for the Travis CI builds, however, formats the grouped filter use statement differently:

    use filters::{abs, append, capitalize, ceil, date, divided_by, downcase, escape, escape_once,
                  first, floor, join, last, lstrip, minus, modulo, newline_to_br, pluralize, plus,
                  prepend, remove, remove_first, replace, replace_first, reverse, round, rstrip, size,
                  slice, sort, split, strip, strip_html, strip_newlines, times, truncate,
                  truncatewords, uniq, upcase};

No big deal, but I figured you'd want to know.

johannhof commented:
> note that the most recent nightly of clippy is complaining about the five uses of lazy_static! (four in lexer.rs, one in filters.rs) so (on my end, at least) the rustup run nightly cargo clippy fails.

Seems like there's an issue around that recently added rule: https://github.com/Manishearth/rust-clippy/issues/1580

You can ignore it by adding `#![cfg_attr(feature="cargo-clippy", allow(zero_ptr))]` to `lib.rs`.

> The rustfmt for the Travis CI builds, however, formats the grouped filter use statement differently:

Feel free to update the version to `v0.7.1-20170120` here: https://github.com/cobalt-org/liquid-rust/blob/master/.travis.yml#L10

johannhof commented:
> You can ignore it by adding #![cfg_attr(feature="cargo-clippy", allow(zero_ptr))] to lib.rs.

Nevermind, I just merged #80 so you'll just have to rebase :)

johannhof commented:
Looks great!
